### PR TITLE
Ensure path writable for MSBuildDebugEngine

### LINKED ISSUE #7241: TypeInitializerException in DebugUtils
### Issue Description

In [this feedback ticket](https://developercommunity.visualstudio.com/t/visual-studio-gets-confused-when-i-run-msbuild/1589332) I advised a customer to set the `MSBuildDebugEngine` environment variable to `1`. When they did so they received...

> error : The type initializer for ‘Microsoft.Build.Shared.Debugging.DebugUtils’ threw an exception.

...for each of the three non-SDK-style .csproj projects in their solution.

### Steps to Reproduce

As above. This is customer reported and I don't have much information at this point.

Feel free to engage the customer via the feedback ticket directly.

### Expected Behavior

No exception.

### Actual Behavior

Exception.

### Analysis

There is quite a bit of type initializer logic in: https://github.com/dotnet/msbuild/blob/518c041f4511a6bc23eb40703b69a94ea46c65fd/src/Shared/Debugging/DebugUtils.cs

I've asked if they have more information in the Event Viewer.

Feel free to engage the customer via the feedback ticket directly.

### Versions & Configurations

Visual Studio 2022 version 17.0.4

### Attach a binlog

This exception means we are blocked on getting build logs.

### PR DESCRIPTION
**Pull request by Forgind**
Fixes #7241

### Context
When MSBuildDebugEngine, it will throw a confusing exception if it finds it cannot write to the specified path. The path is, by default, MSBUILDDEBUGPATH or the current working directory.

### Changes Made
If  MSBUILDDEBUGPATH is invalid, it will now attempt to redirect to the current working directory. If writing to the current working directory fails, perhaps because it's in C:\Program Files or another protected directory, it redirects to TEMP. If the path is too long, it now fails with a nice error.

### Testing
Manually tested to ensure redirection happened for both the argument exception (invalid path characters in MSBUILDDEBUGPATH) and the UnauthorizedAccessException (writing to C:\Program Files) cases.

### CODE REVIEW

**Comment by rainersigwald** [comment]
> If MSBUILDDEBUGPATH is invalid, it will now attempt to redirect to the current working directory. If writing to the current working directory fails, perhaps because it's in C:\Program Files or another protected directory, it redirects to TEMP.

This is very complicated. Can we instead just fail?

**Comment by Forgind** [comment]
> > If MSBUILDDEBUGPATH is invalid, it will now attempt to redirect to the current working directory. If writing to the current working directory fails, perhaps because it's in C:\Program Files or another protected directory, it redirects to TEMP.
> 
> This is very complicated. Can we instead just fail?

Put more simply, we try MSBUILDDEBUGPATH > CWD > TEMP in that order.

We obviously can't take MSBUILDDEBUGPATH out of that list of options. Since it currently defaults to CWD, as do several of our other logs, I don't think we should take that out. The exception that led us to think of investigating this was in which MSBUILDDEBUGPATH was null, and CWD was protected, hence falling back to TEMP. We could just fail, but if we have a good (and easy) alternative, I think it's better to do that first.

File: src/Shared/Debugging/DebugUtils.cs [hunk 1]
```diff
@@ -22,13 +23,27 @@ private enum NodeMode
...
-            string environmentDebugPath = Environment.GetEnvironmentVariable("MSBUILDDEBUGPATH");
-            var debugDirectory = environmentDebugPath;
+            string environmentDebugPath = FileUtilities.TrimAndStripAnyQuotes(Environment.GetEnvironmentVariable("MSBUILDDEBUGPATH"));
+            string debugDirectory = environmentDebugPath;
 
             if (Traits.Instance.DebugEngine)
             {
                 debugDirectory ??= Path.Combine(Directory.GetCurrentDirectory(), "MSBuild_Logs");
 
+                // Probe writeability
```
Comment (BenVillalobos) [inline src/Shared/Debugging/DebugUtils.cs:33]: nit: comment doesn't help much

File: src/Shared/Debugging/DebugUtils.cs [hunk 2]
```diff
@@ -22,13 +23,27 @@ private enum NodeMode
...
+                    debugDirectory = ProbeWriteability(debugDirectory);
+                }
+                catch (ArgumentException)
+                {
+                    // This can happen if MSBUILDDEBUGPATH contains invalid characters, but the current working directory may still work.
+                    debugDirectory = ProbeWriteability(Path.Combine(Directory.GetCurrentDirectory(), "MSBuild_Logs"));
+                }
+
+                // Redirect to TEMP if we failed to write to either MSBUILDDEBUGPATH or the current working directory.
+                debugDirectory ??= Path.Combine(Path.GetTempPath(), "MSBuild_Logs");
```
Comment (BenVillalobos) [inline src/Shared/Debugging/DebugUtils.cs:45]:
```suggestion
                debugDirectory = ProbeWriteability(debugDirectory);
                debugDirectory ??= ProbeWriteability(Path.Combine(Directory.GetCurrentDirectory(), "MSBuild_Logs"));
                debugDirectory ??= Path.Combine(Path.GetTempPath(), "MSBuild_Logs");
```

If `ProbeWriteability` can be changed to catch its own ArgumentException and return null in that case, it simplifies the logic here.

File: src/Shared/Debugging/DebugUtils.cs [hunk 3]
```diff
@@ -44,6 +59,32 @@ static DebugUtils()
             DebugPath = debugDirectory;
         }
 
+        private static string ProbeWriteability(string path)
```
Comment (BenVillalobos) [inline src/Shared/Debugging/DebugUtils.cs:62]: I don't like the naming here. Something like `CanWriteToDirectory` or similar is more intuitive.

See my comment above, but this function can catch its own `ArgumentException` and return null in that case.

We can consider placing this in `FileUtilities` if we can generalize this.

File: src/Shared/FileUtilities.cs [hunk 4]
```diff
@@ -270,6 +270,11 @@ internal static bool IsSlash(char c)
         /// </summary>
         internal static string TrimAndStripAnyQuotes(string path)
         {
+            if (path is null)
```
Comment (BenVillalobos) [inline src/Shared/FileUtilities.cs:273]: 👍!

File: src/Shared/Resources/Strings.shared.resx [hunk 5]
```diff
@@ -278,10 +278,14 @@
     <value>MSB5028: Solution filter file at "{0}" includes project "{1}" that is not in the solution file at "{2}".</value>
     <comment>{StrBegin="MSB5028: "}UE: The project filename is provided separately to loggers.</comment>
   </data>
+  <data name="DebugPathTooLong" Visibility="Public">
+    <value>MSB5029: The path "{0}" used for debug logs is too long. Set it to a shorter value using the MSBUILDDEBUGPATH environment variable or change your system configuration to allow long paths.</value>
```
Comment (BenVillalobos) [inline src/Shared/Resources/Strings.shared.resx:282]: If we mention changing the system to allow long paths, should we also include a help link?

**Review by BenVillalobos — COMMENTED** [review]
> This is very complicated. Can we instead just fail?

I can see skipping the current directory option, but temp could be very useful.

File: src/Shared/Debugging/DebugUtils.cs (hunk 2 above)
Comment (Forgind) [reply src/Shared/Debugging/DebugUtils.cs:45]: I was trying to get out better perf, but on reflection, that isn't a very satisfying reason for making this more complicated. This runs once, and the difference is tiny, so I really shouldn't care. I'll change this.

File: src/Shared/Debugging/DebugUtils.cs (hunk 3 above)
Comment (Forgind) [reply src/Shared/Debugging/DebugUtils.cs:62]: I think the general form should just return true or false for can write/can't write. Should be doable.

File: src/Shared/Resources/Strings.shared.resx (hunk 5 above)
Comment (Forgind) [reply src/Shared/Resources/Strings.shared.resx:282]: Reasonable, but that's a very google-able not-MSBuild thing, so I'd vote no, personally.

File: src/Shared/Debugging/DebugUtils.cs [hunk 6]
```diff
@@ -22,12 +23,23 @@ private enum NodeMode
...
+                    // Debug directory is writable; no need for fallbacks
+                }
+                else if (FileUtilities.CanWriteToDirectory(Directory.GetCurrentDirectory()))
+                {
+                    debugDirectory = Path.Combine(Directory.GetCurrentDirectory(), "MSBuild_Logs");
+                }
+                else
+                {
+                    debugDirectory = Path.Combine(Path.GetTempPath(), "MSBuild_Logs");
+                }
```
Comment (Forgind) [inline src/Shared/Debugging/DebugUtils.cs:42]: 20-30 minutes: figure out if I can log a warning in a visible way in VS

If yes, log warning for these with new path included.

If not, check what happens if I log a catastrophic failure.

**Comment by Forgind** [comment]
Update: After talking with baronfel, we're keeping the current fallback scheme, but I will try to writing to the windows event log and, if that proves too difficult, abandoning writing a message to the user.

**Comment by rainersigwald** [comment]
> I will try to writing to the windows event log and, if that proves too difficult, abandoning writing a message to the user.

I don't understand this as a goal. Why introduce a new logging mechanism for MSBuild here?

**Comment by baronfel** [comment]
I'm not married to it, but systems have places like the EventLog (on Windows) and Syslog (on Mac) as an ever-present last-change logging target for a reason.  In the hosted scenario (VS/VSMac) we don't have full control over the logger config, so if we do want to provide some sort of logging to the user (in a way that doesn't cause an application exception at startup), those mechanisms are viable IMO. Especially for a developer-centric audience, where the EventLog/Syslog is a known _thing_. 

That being said, to add this capability in a complete way (i.e an OS-agnostic interface with OS-specific implementations) might be too much of an ask for this PR, so I'd be willing to let it go.

**Comment by Forgind** [comment]
I tested out my code, and it threw a SecurityException, which from what I can tell online, means I didn't run it in admin mode, and since there wasn't already an MSBuild EventLog key, it failed. Maybe we can add an MSBuild EventLog key in setup? But I'm not sure it's worth it. We could theoretically log under Application or .NET Runtime, but neither of those feels quite right. Thoughts?

**Comment by baronfel** [comment]
That's enough of a roadblock to throw away the idea from my perspective @Forgind.

**Comment by Forgind** [comment]
I found a way to log a deferred message. It isn't a warning, but it's a high-pri message if MSBuildDebugPath was not set or invalid and a low-pri message if it was.

File: src/Shared/Debugging/DebugUtils.cs [hunk 7]
```diff
@@ -22,12 +23,27 @@ private enum NodeMode
...
+            DebugPathMessageImportance = MessageImportance.Low;
 
             if (Traits.Instance.DebugEngine)
             {
-                debugDirectory ??= Path.Combine(Directory.GetCurrentDirectory(), "MSBuild_Logs");
+                DebugPathMessageImportance = MessageImportance.High;
+                if (!string.IsNullOrWhiteSpace(debugDirectory) && FileUtilities.CanWriteToDirectory(debugDirectory))
+                {
+                    // Debug directory is writable; no need for fallbacks
+                    DebugPathMessageImportance = MessageImportance.Low;
```
Comment (rainersigwald) [inline src/Shared/Debugging/DebugUtils.cs:37]: I think I'd prefer to leave this as high: it's interesting that they've used the env var to set up logging even if it works.

File: src/Shared/FileUtilities.cs [hunk 8]
```diff
@@ -165,6 +165,30 @@ internal static int GetPathsHash(IEnumerable<string> assemblyPaths)
...
+        /// <summary>
+        /// Returns whether MSBuild can write to the given directory. Throws for PathTooLongExceptions
+        /// but not other exceptions.
+        /// </summary>
+        internal static bool CanWriteToDirectory(string directory)
+        {
+            try
+            {
+                string testFilePath = Path.Combine(directory, "textFile.txt");
+                File.WriteAllText(testFilePath, "Successfully wrote to file.");
```
Comment (rainersigwald) [inline src/Shared/FileUtilities.cs:177]: With some tweaks it'll be much easier to determine if this is causing problems, and support concurrent execution:

``suggestion
                string testFilePath = Path.Combine(directory, $"MSBuild_{Guid.NewGuid().ToString("N")}.txt");                File.WriteAllText(testFilePath, $"MSBuild process {Process.GetCurrentProcess().Id} successfully wrote to file.");
```

File: src/MSBuild/XMake.cs [hunk 9]
```diff
@@ -1365,12 +1365,19 @@ private static IEnumerable<BuildManager.DeferredBuildMessage> GetMessagesToLogIn
                         "MSBVersion",
                         ProjectCollection.DisplayVersion),
                     MessageImportance.Low),
-                new BuildManager.DeferredBuildMessage(
-                    ResourceUtilities.FormatResourceStringIgnoreCodeAndKeyword(
+            };
+
+            if (!string.IsNullOrEmpty(Environment.GetEnvironmentVariable("MSBuildDebugEngine")))
```
Comment (rainersigwald) [inline src/MSBuild/XMake.cs:1370]: Would it be better to check `DebugUtils.DebugPath is not null`, so the env var stuff only has to happen once?

**Review by rainersigwald — APPROVED** [review]

File: src/MSBuild/XMake.cs (hunk 9 above)
Comment (Forgind) [reply src/MSBuild/XMake.cs:1370]: I thought about that. DebugUtils.DebugPath is not null if MSBUILDDEBUGPATH is set even if MSBuildDebugEngine is not enabled, which I didn't like. Then it isn't a VS thing but also a thing people (or at least I) really do use from the command line, and a high priority message about a debug path that I know where is felt undesirable. That said, I should be able to reference the trait and get the best of both worlds:
```suggestion
            if (Traits.Instance.DebugEngine)
```

That's only risky if someone changes the environment variable at an awkward time, which shouldn't be important.

**Pull request merged**